keras_nlp/models/bert.py

```diff
@@ -14,128 +14,218 @@
-
-    This helper will validate the `weights` and `vocabulary_size` arguments, and
-    fully resolve them in the case we are loading pretrained weights.
-    """
-    if (vocabulary_size is None and weights is None) or (
-        vocabulary_size and weights
-    ):
-        raise ValueError(
-            "One of `vocabulary_size` or `weights` must be specified "
-            "(but not both). "
-            f"Received: weights={weights}, "
-            f"vocabulary_size={vocabulary_size}"
-        )
-
-    if weights:
-        if weights not in checkpoints[bert_variant]:
-            raise ValueError(
-                "`weights` must be one of "
-                f"""{", ".join(checkpoints[bert_variant])}. """
-                f"Received: {weights}"
-            )
-
-        vocabulary_size = checkpoints[bert_variant][weights]["vocabulary_size"]
-
-        # TODO(jbischof): consider changing format from `h5` to
-        # `tf.train.Checkpoint` once
-        # https://github.com/keras-team/keras/issues/16946 is resolved.
-        weights = keras.utils.get_file(
-            "model.h5",
-            BASE_PATH + f"{bert_variant}_{weights}/model.h5/",
-            cache_subdir=f"models/{bert_variant}/{weights}/",
-            file_hash=checkpoints[bert_variant][weights]["md5"],
-        )
-
-    return weights, vocabulary_size
-
-
-# Pretrained models
-BASE_PATH = "https://storage.googleapis.com/keras-nlp/models/"
-
+# Metadata for loading pretrained model weights.
+# We need the vocabulary_size hardcoded so we can instantiate a Bert network
+# with the right embedding size without downloading the matching vocabulary.
 checkpoints = {
     "bert_tiny": {
         "uncased_en": {
-            "md5": "c2b29fcbf8f814a0812e4ab89ef5c068",
-            "description": "Tiny size of BERT where all input is lowercased. "
-            "Trained on English Wikipedia + BooksCorpus.",
+            "description": (
+                "Tiny size of BERT where all input is lowercased. "
+                "Trained on English Wikipedia + BooksCorpus."
+            ),
+            "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_tiny_uncased_en/model.h5",
+            "weights_hash": "c2b29fcbf8f814a0812e4ab89ef5c068",
             "vocabulary_size": 30522,
         }
     },
     "bert_small": {
         "uncased_en": {
-            "md5": "08632c9479b034f342ba2c2b7afba5f7",
-            "description": "Small size of BERT where all input is lowercased. "
-            "Trained on English Wikipedia + BooksCorpus.",
+            "description": (
+                "Small size of BERT where all input is lowercased. "
+                "Trained on English Wikipedia + BooksCorpus."
+            ),
+            "weights_url": "https://storage.googleapis.com/keras-nlp/models/bert_small_uncased_en/model.h5",
+            "weights_hash": "08632c9479b034f342ba2c2b7afba5f7",
             "vocabulary_size": 30522,
```

Comment:
Maybe this is out of scope, but in the longer term it seems like vocabularies should know their size and pretrained models should point to a vocabulary. The current setup doesn't make a ton of sense.

Reply:
That makes sense to me. Will mean that the graph code dips into both of these metadata dicts, but I think that kinda is the coupling here right? Will try that out.